### ISSUE TITLE
Staker proxy contract.

### ISSUE THREAD

**Issue by deepesh-kn** [OpenST/mosaic-contracts#741]
This is closely dependent on #740 

**StakerProxy contract:**

**Please note:** the contract is referred to `StakerProxy`, but this is not final name. Suggest a better name.

Gateways can do one stake - mint process at a time for a given staker address. So the composer contract cannot call gateway.stake() function, it will be not able to serve all stake requests from multiple stakers addresses. So a proxy contract is introduced as a proposal here.

- This contract has a minimal code, that gets deployed from the `Composer` contract.
- It is deployed per user once.
- It can be used for arbitrary gateways.
- Actual staker is the owner, and it can be destroyed by the owner only.
- All other functions can be called only by Composer contract.

_______________________________
This contract can have following variables.
```
address  public composer;
address  public owner;
```
Constructor
```  
    constructor(
        address _composer, 
        address _owner
    ) 
        public;
```
Accept stake request
```
     function acceptStakeRequest(
        uint256 _amount,
        address _beneficiary,
        uint256 _gasPrice,
        uint256 _gasLimit,
        uint256 _nonce,
        bytes32 _hashLock,
        address _staker,
        address _gateway
    )
        onlyComposer
        external
        returns (bytes32 messageHash_);
```

More detail is on EPIC

**Comment by abhayks1** [comment]
Was going through the epic 📔 . Have a question 

- Should we also support `transfer` and `approve` methods in the proxy contract so that transferred money is not stuck? This could happen because of revertStake(if we are supporting) or an arbitrary transfer.  

@deepesh-kn @schemar what do you think 🤔 ?

**Comment by deepesh-kn** [comment]
> Was going through the epic 📔 . Have a question
> 
> * Should we also support `transfer` and `approve` methods in the proxy contract so that transferred money is not stuck? This could happen because of revertStake(if we are supporting) or an arbitrary transfer.
> 
> @deepesh-kn @schemar what do you think 🤔 ?

Nice point 👍 
This is kept out of scope for now. We can think about this.

Revert will be very tricky. A facilitator will not get any benefit for revert stake, as the bounty will be burnt and also will not get any reward for facilitation so it may not do the facilitation. In this case staker needs to do facilitation itself. Staker may not have OST' on the auxiliary chain to perform it.

In case the tokens are stuck in the gateway then, any facilitator can progress with proof, the facilitator can get bounty and reward.

**Comment by abhayks1** [comment]
`In case the tokens are stuck in the gateway then, any facilitator can progress with proof, the facilitator can get bounty and reward.`

I meant if tokens are stuck in staker proxy address. An arbitrary transfer(e.g. ALT coins) can be made to staker proxy address since the address is public. Should we give owner option to take out that money?

**Comment by schemar** [comment]
Right now I added a self-destruct method that returns all base coin and erc20 value to the owner before destructing.

**Comment by deepesh-kn** [comment]
> Right now I added a self-destruct method that returns all base coin and erc20 value to the owner before destructing.

It should also update the composer contract to remove the proxy contract address for the staker address? We can discuss this in detail.

**Comment by schemar** [comment]
We had a short call and discussed that we want to:

* use the proxy for any number of gateways
* add `nonce` and `gateway` to stake method; nonce only for checking
* add a mutex as you could provide a malicious gateway
* add a `transferToken` function to get funds out of the proxy (e.g. after revert)
* remove all transfers from `selfDestruct` (they are unknown for any number of gateways)
* add a comment on self destruct that staking value (incl. bounty) is not stored on the contract, only "passed through".
* add a method to the composer to remove a proxy on self destruct (method name pending)
